src/tablenavigation.js

```diff
@@ -0,0 +1,515 @@
+		const editing = this.editor.editing;
+		const domConverter = editing.view.domConverter;
+
+		// Wrapped lines contain exactly the same position at the end of current line
+		// and at the beginning of next line. That position's client rect is at the end
+		// of current line. In case of caret at first position of the last line that 'dual'
+		// position would be detected as it's not the last line.
+		if ( isForward ) {
+			const probe = model.createSelection( modelRange.start );
+
+			model.modifySelection( probe );
+
+			// If the new position is at the end of the container then we can't use this position
+			// because it would provide incorrect result for eg caption of image and selection
+			// just before end of it. Also in this case there is no "dual" position.
+			if ( !probe.focus.isAtEnd && !modelRange.start.isEqual( probe.focus ) ) {
+				modelRange = new ModelRange( probe.focus, modelRange.end );
+			}
+		}
+
+		const viewRange = editing.mapper.toViewRange( modelRange );
+		const domRange = domConverter.viewRangeToDom( viewRange );
+		const rects = Rect.getDomRangeRects( domRange );
+
+		let boundaryVerticalPosition;
+
+		for ( const rect of rects ) {
+			if ( boundaryVerticalPosition === undefined ) {
+				boundaryVerticalPosition = Math.round( rect.bottom );
+				continue;
+			}
+
+			// Let's check if this rect is in new line.
+			if ( Math.round( rect.top ) >= boundaryVerticalPosition ) {
+				return false;
+			}
+
+			boundaryVerticalPosition = Math.max( boundaryVerticalPosition, Math.round( rect.bottom ) );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Moves the selection from the given table cell in the specified direction.
+	 *
+	 * @private
+	 * @param {module:engine/model/element~Element} tableCell The table cell to start the selection navigation.
+	 * @param {'left'|'up'|'right'|'down'} direction Direction in which selection should move.
+	 */
+	_navigateFromCellInDirection( tableCell, direction ) {
+		const model = this.editor.model;
+
+		const table = findAncestor( 'table', tableCell );
+		const tableMap = [ ...new TableWalker( table, { includeSpanned: true } ) ];
+		const { row: lastRow, column: lastColumn } = tableMap[ tableMap.length - 1 ];
+
+		const currentCellInfo = tableMap.find( ( { cell } ) => cell == tableCell );
+		let { row, column } = currentCellInfo;
+
+		switch ( direction ) {
+			case 'left':
+				column--;
+				break;
+
+			case 'up':
+				row--;
+				break;
+
+			case 'right':
+				column += currentCellInfo.colspan;
+				break;
+
+			case 'down':
+				row += currentCellInfo.rowspan;
+				break;
+		}
+
+		const isOutsideVertically = row < 0 || row > lastRow;
+		const isBeforeFirstCell = column < 0 && row <= 0;
+		const isAfterLastCell = column > lastColumn && row >= lastRow;
+
+		// Note that if the last table cell is row-spanned then isAfterLastCell will never be true but we don't know
+		// if user was navigating on the last row or not, so let's allow him to stay in the table.
```

Comment:
Do you have test for this? I think that this is not possible - we fix table if all cells have bigger rowspan than "visual" one.

Ie create a 2x4 table and merge all cells in the first column (it will have rowspan=4) then remove the last row. The rowspan will be reduced to 3.

Reply:
This comment was describing this situation:
![Apr-22-2020 20-19-40](https://user-images.githubusercontent.com/1232187/80018729-c28b5380-84d6-11ea-98fd-d3cdc0dab0ed.gif)
This is the same behavior as in GDocs.

Reply:
```suggestion
		// Note that if the table cell at the end of a row is row-spanned then isAfterLastCell will never be true.
		// However, we don't know if user was navigating on the last row or not, so let's stay in the table.
```